Adjust buffering in examples

ajfrantz opened the pull request:
## Motivation

#2444 identified an example (`echo.rs`) which uses a stack-based buffer and suggests it would be improved by using a heap-based buffer instead.

## Solution

`echo.rs` is a straightforward switch from array to `Vec`. `BytesMut` was also suggested in discussion on #2444, but I chose `Vec` because it matches the approach in [the website tutorial](https://tokio.rs/tokio/tutorial/io).

Darksonn commented on examples/connect.rs:
```diff
@@ -112,36 +113,26 @@ mod udp {
-        mut stdout: impl Sink<Bytes, Error = io::Error> + Unpin,
-        reader: &UdpSocket,
-    ) -> Result<(), io::Error> {
-        loop {
-            let mut buf = vec![0; 1024];
```
You can avoid multiple allocations here by using an `BytesMut` and splitting off a `Bytes` in each iteration.

Darksonn commented on examples/connect.rs:
```diff
@@ -112,36 +113,26 @@ mod udp {
+            })
+        });
 
+        future::try_join(sink.send_all(&mut stdin), stdout.send_all(&mut stream)).await?;
```
These should be updated to `tokio::try_join!`.

Darksonn commented:
I would prefer to use the existing strategy in `connect.rs`. I think the alternative is more complex than needs be.

ajfrantz commented:
I reverted the `connect.rs` changes. My intention had been to avoid per-packet allocations entirely, however upon further inspection, both the `UdpFramed` version and a theoretical `BytesMut` version fail to achieve that goal: unless I'm missing something, since `stdout` is `Sink<Bytes...`, any data split from a `BytesMut` must be frozen into a `Bytes`, and subsequently can't be `unspilt` back into the original buffer. This means the `BytesMut` will eventually be exhausted and re-allocated, so we might as well do it on every iteration and hope we keep getting the same memory back.

The `echo.rs` change referred to in the issue remain, and I also updated the `try_join` calls throughout the examples, as noted in the review.
